### DISCUSSION TITLE
Joining as colaborator.

### DISCUSSION THREAD

**Discussion by urosjarc** [pfusik/cito#75] [Q&A]
Hy @pfusik and all admins of this repo. I came across this repo because I want to create 50+ exercises for my students in various languages (the first year is python, the second year is java, the third year is cpp). I want to create empty templates with provided tests already in the templates so that my students can check if they made the correct implementation.

Currently, I'm in testing mode to check if this project is in a viable state to be used in my goals.
I have checked [haxe programming language](https://haxe.org/) that have similar capabilities but the boilerplate code used in their transpiler is not a viable option for my easily scared students.

I'm pretty sure that I will come across many issues and I want to be able to create pull requests with the proper fixes and tests so that my goals will be accomplished.

Is there any option to arange a zoom metup in which the code and functionality of this project be explained to me, so that I can start fixing bugs. Zoom meeting can be record and published on youtube for future enthusiasts wanting to be a part of this project.

What do you guys think about that?

**Comment by pfusik** [comment]
Welcome Uroš!

Currently `cito` is a one-man project. :)

I haven't thought of using it as a teaching tool. But I guess it might work. Other transpilers, such as Nim, emit tons of boilerplate code, while Ć aims to emit exactly what an experienced programmer would write.

What kind of tests are you interested in? If you want to use the target languages' frameworks, that's not implemented, but there's [an open ticket](https://github.com/pfusik/cito/issues/44).

I'll be happy to merge good PRs. For bigger features, please discuss first.

Yes, we can meet on Zoom next week. This week I'm traveling. Please write an agenda.

**Reply by urosjarc** [reply]
The agenda (expected time 2h) would be like this...

1. To go over the makefile to clarify some things there because I don't understand some things in that file.
2. How do tests work in this project...
3. Explain basic arhitecture how classes are connected in class diagram and their relationships.
4. To go over every class in the project and explain basic logic behind code in every class function. 
5. Explain recomendet workflow when implementing new feature.
6. Explain what are some issues in the project that are not resolved yet.

I will record our talk and upload the video to youtube with your approval if it's alright with you and provide link in cito readme.
Please can you create new google event when its alright with you to meet... My email is jar.fmf@gmail.com

<hr>

How I would tests students code goes something like this... I will write base templates in C# editor Rider (jetbrains)

## I provide student with code templat
```
/// <summary>
/// Številski sistem predstavlja način s katerim lahko zapišemo številke.
/// V tem odseku se bomo naučili kako pretvarjati številke v različne formate
/// ki se uporabljajo v programiranju.
/// <remarks>number, string</remarks>
/// </summary>
public class number_system
{
    /// <summary>
    /// Pretvori zapis številke iz desetiškega v binarnega.
    /// </summary>
    /// <param name="a">Cela številka v desetiškem zapisu.</param>
    /// <returns>Številko <param name="a"></param> v binarnem zapisu.</returns>
    public static int decimal_to_binary(int a)
    {
        return 0;
    }

    /// <summary>
    /// Pretvori zapis številke iz desetiškega v šestnajstinskega.
    /// </summary>
    /// <param name="a">Cela številka v desetiškem zapisu.</param>
    /// <returns>Številka <param name="a"></param> v šestnajstinskem zapisu.</returns>
    public static string decimal_to_hex(int a)
    {
        return "";
    }

    static void Main(string[] args)
    {
        Console.Out.WriteLine($"\n   {decimal_to_binary(10) == 1010} -> decimal_to_binary");
        Console.Out.WriteLine($"   {decimal_to_hex(10) == "a0f"} -> decimal_to_hex");
    }
}

```

## Student create function implementation

They can check if their implementation is somewhat correct with simple tests in the main.
But when they want to submit results they will have to create server request on my server.
I will provide them with the python file witch will fetch tests data for specific function from the server
and test local function of the student... Then I will send back to the server function results, server then returns
which tests passed... with that I will monitor students progress how they progress.

As with the cito tests in #44 this is impossible task because languages like js do not have an official supported test framework as python has.

**Reply by urosjarc** [reply]
@pfusik After deep consideration I have decided to create my own syntax tree compiler in python since empty templates does not take to much work to be created for every language (I almost finished 80% of the work). Working on cito would be for me to much of time burden. All the best!

**Reply by pfusik** [reply]
Okay. Good luck!
